### PR TITLE
Support Deconvolution3D on IE backend

### PR DESCRIPTION
**Pull request by l-bat**
**Merge with extra:** opencv/opencv_extra#632

```
force_builders=Custom,Custom Win,Custom Mac
build_image:Custom=ubuntu-openvino-2019r1:16.04
build_image:Custom Win=openvino-2019r1
build_image:Custom Mac=openvino-2019r1

test_modules:Custom=dnn,python2,python3,java
test_modules:Custom Win=dnn,python2,python3,java
test_modules:Custom Mac=dnn,python2,python3,java

buildworker:Custom=linux-1
test_opencl:Custom=ON
test_bigdata:Custom=1
test_filter:Custom=*
```

### CODE REVIEW

File: modules/dnn/src/layers/convolution_layer.cpp [hunk 1]
```diff
@@ -1764,7 +1789,7 @@ class DeConvolutionLayerImpl CV_FINAL : public BaseConvolutionLayerImpl
             transpose(weightsMat, newWeights);
         }
 
-        const int outGroupCn = blobs[0].size[1];  // Weights are in IOHW layout
+        const int outGroupCn = blobs[0].size[1];  // Weights are in IOHW or NCDHW layout
```
Comment (dkurt) [inline modules/dnn/src/layers/convolution_layer.cpp]: `IODHW`?

File: modules/dnn/test/test_onnx_importer.cpp [hunk 2]
```diff
@@ -119,6 +119,19 @@ TEST_P(Test_ONNX_layers, Deconvolution)
     testONNXModels("deconv_adjpad_2d", npy, 0, 0, false, false);
 }
 
+TEST_P(Test_ONNX_layers, Deconvolution3D)
+{
+#if defined(INF_ENGINE_RELEASE) && INF_ENGINE_VER_MAJOR_LT(2019010000)
+    throw SkipTestException("Test is enabled starts from 2019R1");
+#endif
+    if (backend != DNN_BACKEND_INFERENCE_ENGINE || target != DNN_TARGET_CPU)
+        throw SkipTestException("Only DLIE backend on CPU is supported");
```
Comment (dkurt) [inline modules/dnn/test/test_onnx_importer.cpp]: Please use `applyTestTag`:

File: modules/dnn/src/layers/convolution_layer.cpp [hunk 3]
```diff
@@ -72,6 +73,21 @@ class BaseConvolutionLayerImpl : public ConvolutionLayer
...
+                adjust_pads.push_back(adj_pad.get<int>(i));
+            }
+        } else {
+            if (kernel_size.size() == 2) {
+                adjust_pads.push_back(params.get<int>("adj_h", 0));
+                adjust_pads.push_back(params.get<int>("adj_w", 0));
+            } else if (kernel_size.size() == 3) {
+                adjust_pads = std::vector<size_t>(3, 0);
+            }
+        }
```
Comment (dkurt) [inline modules/dnn/src/layers/convolution_layer.cpp]: Is it possible to use `getParameter` here as we do for kernels and strides? Perhaps, this way we need to rename `adj_pad` to just `adj`.

File: modules/dnn/test/test_onnx_importer.cpp (hunk 2 above)
Comment (dkurt) [reply modules/dnn/test/test_onnx_importer.cpp]: Please do the same for second skip

File: modules/dnn/src/layers/convolution_layer.cpp [hunk 4]
```diff
@@ -1136,24 +1137,34 @@ class DeConvolutionLayerImpl CV_FINAL : public BaseConvolutionLayerImpl
...
+                return false;
+            }
 
-            if (adjustPad.height || adjustPad.width)
+            if (std::accumulate(adjust_pads.begin(), adjust_pads.end(), 1, std::plus<size_t>()) > 0)
             {
                 if (padMode.empty())
                 {
-                    if (preferableTarget != DNN_TARGET_CPU && group != 1)
+                    if ((kernel_size.size() == 3 || preferableTarget != DNN_TARGET_CPU) && group != 1)
```
Comment (dkurt) [inline modules/dnn/src/layers/convolution_layer.cpp]: Why we need to check `kernel_size.size() == 3` only?

File: modules/dnn/src/onnx/onnx_importer.cpp [hunk 5]
```diff
@@ -645,42 +645,33 @@ void ONNXImporter::populateNet(Net dstNet)
...
-                    DictValue kernel = layerParams.get("kernel_size");
-                    layerParams.set("adj_h", (outH - kernel.getIntValue(0)) % strideY);
-                    layerParams.set("adj_w", (outW - kernel.getIntValue(1)) % strideX);
+                    padMode = toUpperCase(layerParams.get<String>("pad_mode"));
+                    if (padMode != "SAME" && padMode != "VALID")
+                        CV_Error(Error::StsError, "Unsupported padding mode " + padMode);
+
+                    for (int i = 0; i < stride.size(); i++)
+                        adjust_pads.push_back(padMode == "SAME" ? (outShape.get<int>(2 + i) - 1) % stride.get<int>(i) :
+                                                                  (outShape.get<int>(2 + i) - kernel.get<int>(i)) % stride.get<int>(i));
```
Comment (dkurt) [inline modules/dnn/src/onnx/onnx_importer.cpp]: Please use separate `int` variables for dimension and stride:

```cpp
int sz = outShape.get<int>(2 + i);
int stride = stride.get<int>(i);
```

**Comment by dkurt** [comment]
Please check failed tests: https://pullrequest.opencv.org/buildbot/builders/precommit_custom_linux/builds/2350

File: modules/dnn/src/layers/convolution_layer.cpp (hunk 4 above)
Comment (l-bat) [reply modules/dnn/src/layers/convolution_layer.cpp]: done

File: modules/dnn/src/onnx/onnx_importer.cpp (hunk 5 above)
Comment (l-bat) [reply modules/dnn/src/onnx/onnx_importer.cpp]: done

**Pull request merged**